### ISSUE TITLE
Regression in loading optimizer learning rate

### ISSUE THREAD

**Issue by Jackmin801** [pytorch/pytorch#129079]
### 🐛 Describe the bug

Hey PyTorch team! Thanks for your hard work on the distributed checkpointing, big fan of the work!

There seems to have been a regression between `2.2.2` and `2.3.0` that resulted in the optimizer learning rate being loaded incorrectly. It seems like the `lr` key in params group is not loaded from the checkpoint, resulting in the first step having the initialisation learning rate instead of the checkpointed one. This was not the case in `2.2.2` and only occurs after `2.3.0`. 
 
 ### debug_dcp.py
 ```python
 import os

import torch
import torch.distributed as dist
import torch.distributed.checkpoint as dcp
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, ShardingStrategy
from torch.distributed.checkpoint.state_dict import get_state_dict, set_state_dict
from transformers import AutoModelForCausalLM, get_cosine_schedule_with_warmup
from cyclopts import App
import hashlib

def setup():
    # initialize the process group
    dist.init_process_group("nccl")
    torch.cuda.set_device(int(os.environ['LOCAL_RANK']))

def cleanup():
    dist.destroy_process_group()

def save_checkpoint(checkpoint_path: str, model: torch.nn.Module, optimizer: torch.optim.Optimizer, scheduler: torch.optim.lr_scheduler.LambdaLR):
    """Save the model and optimizer state to a checkpoint folder

    Args:
        checkpoint_path: the path to the checkpoint folder
        model: the model to save
        optimizer: the optimizer to save
        scheduler: the scheduler to save
    """

    # 1. Save distributed states
    fs_storage_writer = dcp.FileSystemWriter(checkpoint_path)

    model_state_dict, optimizer_state_dict = get_state_dict(model, optimizer)
    dcp_state_dict = {
        "model": model_state_dict,
        "optimizer": optimizer_state_dict,
    }
    dcp.save(dcp_state_dict, storage_writer=fs_storage_writer)

    # 2. Save global states
    global_state_dict = {
        "scheduler": scheduler.state_dict(),
    }
    torch.save(global_state_dict, os.path.join(checkpoint_path, "global_state_dict.pt"))

def load_checkpoint(checkpoint_path: str, model: torch.nn.Module, optimizer: torch.optim.Optimizer, scheduler: torch.optim.lr_scheduler.LambdaLR) -> dict:
    """Load the model and optimizer state from a checkpoint folder
    
    Args:
        checkpoint_path: the path to the checkpoint folder
        model: the model to load
        optimizer: the optimizer to load
        scheduler: the scheduler to load
    """
    # 1. Load distributed states
    fs_storage_reader = dcp.FileSystemReader(checkpoint_path)

    model_state_dict, optimizer_state_dict = get_state_dict(model, optimizer)
    dcp_state_dict = {
        "model": model_state_dict,
        "optimizer": optimizer_state_dict,
    }
    dcp.load(dcp_state_dict, storage_reader=fs_storage_reader)
    set_state_dict(model, optimizer, model_state_dict=model_state_dict, optim_state_dict=optimizer_state_dict)
    
    # 2. Load global states
    global_state_dict = torch.load(os.path.join(checkpoint_path, "global_state_dict.pt"))
    scheduler.load_state_dict(global_state_dict["scheduler"])

def _round_str(x: float):
    return f"{x:.4f}"

def _round_flatten(a: torch.Tensor, max_size: int = 1000):
    bounds = int(max_size**0.5)
    if len(a.shape) == 1:
        return ",".join(_round_str(i) for i in a[:bounds].flatten())
    elif len(a.shape) == 2:
        return ",".join(_round_str(i) for i in a[:bounds, :bounds].flatten())

def hash_tensor_content(a: torch.Tensor, max_size: int = 1000) -> str:
    return hashlib.md5(_round_flatten(a, max_size=max_size).encode("utf-8")).hexdigest()


app = App()

@app.default
def main(
    save_path: str | None = None,
    load_path: str | None = None,
    ckpt_interval: int = 5,
    total_steps: int = 10,
    trace_file: str | None = None,
):
    rank = int(os.environ['RANK'])
    world_size = int(os.environ['WORLD_SIZE'])
    print(f"Running basic FSDP checkpoint saving example on rank {rank}.")

    # create a model and move it to GPU with id rank
    model = AutoModelForCausalLM.from_pretrained("gpt2").to(rank)
    model = FSDP(model, use_orig_params=False, sharding_strategy=ShardingStrategy.NO_SHARD)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=10, num_training_steps=100)
    if load_path is not None:
        load_checkpoint(load_path, model, optimizer, scheduler)

    for i in range(total_steps):
        input = torch.arange(16).reshape(2, 8).to(rank)

        print(f"[Rank {rank}] Step {i}", hash_tensor_content(next(model.parameters()).data), flush=True)
        #optimizer.param_groups[0]["lr"] = scheduler.get_last_lr()[0]
        output = model(input, labels=input)
        print(output.loss)

        if trace_file is not None and rank == 0:
            with open(trace_file, "a") as f:
                f.write(f"Rank {rank} Step {i} {hash_tensor_content(next(model.parameters()).data)}\n")

        optimizer.zero_grad()
        output.loss.backward()
        print(f"LR before step: {optimizer.param_groups[0]['lr']}")
        optimizer.step()
        scheduler.step()

        if save_path is not None and (i + 1) % ckpt_interval == 0:
            save_checkpoint(save_path + f"/step_{i}", model, optimizer, scheduler)

if __name__ == "__main__":
    setup()
    app()
    cleanup()
 ```
 
# Torch 2.2.2
 ```
 torchrun --nproc-per-node 2 debug_dcp.py --save-path meow2.2-ckpts --trace-file scratch-222.txt
 torchrun --nproc-per-node 2 debug_dcp.py --load-path meow2.2-ckpts/step_4 --trace-file cont-222.txt --total-steps 5
 ```
 
 ### scratch-222.txt
 ```
Rank 0 Step 0 3b5d688fe763cae43f622df6f8302f3f
Rank 0 Step 1 3b5d688fe763cae43f622df6f8302f3f
Rank 0 Step 2 c24c79a4809d91ac9f040f64b4b827c1
Rank 0 Step 3 0ef790d5045b6fc8bce4799cc45364b8
Rank 0 Step 4 963b50b549da0b713bf0add0f71ab02f
Rank 0 Step 5 8d50a54371bfd240829260e16618e7d2
Rank 0 Step 6 192eb2bc4b66a25e8792f0e19302381a
Rank 0 Step 7 4c80ef0da32a7f4407cd7e55341472a9
Rank 0 Step 8 f3a258b46c63a69c65853297bce2714a
Rank 0 Step 9 7d4b62d768ac1ef8d7da9c10d01f4b0a
 ```

### cont-222.txt
```
Rank 0 Step 0 8d50a54371bfd240829260e16618e7d2
Rank 0 Step 1 192eb2bc4b66a25e8792f0e19302381a
Rank 0 Step 2 4c80ef0da32a7f4407cd7e55341472a9
Rank 0 Step 3 f3a258b46c63a69c65853297bce2714a
Rank 0 Step 4 7d4b62d768ac1ef8d7da9c10d01f4b0a
``` 

# Torch 2.3.0
 ```
 torchrun --nproc-per-node 2 debug_dcp.py --save-path meow2.3-ckpts --trace-file scratch-230.txt
 torchrun --nproc-per-node 2 debug_dcp.py --load-path meow2.3-ckpts/step_4 --trace-file cont-230.txt --total-steps 5
 ```

### scratch-230.txt
```
Rank 0 Step 0 3b5d688fe763cae43f622df6f8302f3f
Rank 0 Step 1 3b5d688fe763cae43f622df6f8302f3f
Rank 0 Step 2 c24c79a4809d91ac9f040f64b4b827c1
Rank 0 Step 3 0ef790d5045b6fc8bce4799cc45364b8
Rank 0 Step 4 963b50b549da0b713bf0add0f71ab02f
Rank 0 Step 5 8d50a54371bfd240829260e16618e7d2
Rank 0 Step 6 192eb2bc4b66a25e8792f0e19302381a
Rank 0 Step 7 4c80ef0da32a7f4407cd7e55341472a9
Rank 0 Step 8 f3a258b46c63a69c65853297bce2714a
Rank 0 Step 9 7d4b62d768ac1ef8d7da9c10d01f4b0a
```

### cont-230.txt
```
Rank 0 Step 0 8d50a54371bfd240829260e16618e7d2
Rank 0 Step 1 8d50a54371bfd240829260e16618e7d2
Rank 0 Step 2 9904ec935a7debbf5a4635f5c461648b
Rank 0 Step 3 90771bbfab7aa3d13784354d4de21997
Rank 0 Step 4 53367e9db77942b7659861e928413360
```
 
 cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC @MeetVadakkanchery @mhorowitz 

### Versions

PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-35-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 535.86.10
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper 2950X 16-Core Processor
CPU family:                         23
Model:                              8
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        3500.0000
CPU min MHz:                        2200.0000
BogoMIPS:                           6999.15
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          512 KiB (16 instances)
L1i cache:                          1 MiB (16 instances)
L2 cache:                           8 MiB (16 instances)
L3 cache:                           32 MiB (4 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] triton==2.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi

**Comment by Jackmin801** [comment]
Did a git bisect and found that `f518cf811d8645b801de773c8d3f44fc00d9af1e` is the first bad commit

**Comment by Jackmin801** [comment]
Did a git bisect and found that f518cf811d8645b801de773c8d3f44fc00d9af1e is the first bad commit
@wz337

**Comment by Jackmin801** [comment]
Narrowed down the root cause to the `tree_map_only` used by `_init_state_dict`.
https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/planner_helpers.py#L297-L303
```python
def _init_state_dict(state_dict: STATE_DICT_TYPE) -> None:
    state_dict_assigned_storage = tree_map_only(
        torch.Tensor, lambda v: _init_meta_tensor(v), state_dict
    )
    # The inplace version of tree_map_only, tree_map_only_ doesn't seem to work.
    # So we need to temporariy update the each element in the state dict with meta tensor.
    for k in state_dict.keys():
        state_dict[k] = state_dict_assigned_storage[k]
 ```

Doing the `tree_map_only` out-of-place and then assigning it this way only maintains references for the leaves but not the branches. Which I guess breaks some assumptions made by the reading code. Using the inplace `tree_map_only_` seems to resolve the issue, but im not sure if it breaks something else as suggested by @wz337 in the comment.

```python
def _init_state_dict(state_dict: STATE_DICT_TYPE) -> None:
    tree_map_only_(torch.Tensor, _init_meta_tensor, state_dict)
 ```

**Comment by fegin** [comment]
@Jackmin801 For now, you can use `dcp_state_dict["optimizer"]` to ensure getting the correct loaded state_dict. @wz337 I think our test cases do not catch this use case where we access the state_dict through the original reference.

**Comment by Jackmin801** [comment]
@fegin It seems like `dcp_state_dict["optimizer"]` doesnt work either.

```python
def load_checkpoint(checkpoint_path: str, model: torch.nn.Module, optimizer: torch.optim.Optimizer, scheduler: torch.optim.lr_scheduler.LambdaLR) -> dict:
   """Load the model and optimizer state from a checkpoint folder
   
   Args:
       checkpoint_path: the path to the checkpoint folder
       model: the model to load
       optimizer: the optimizer to load
       scheduler: the scheduler to load
   """
   # 1. Load distributed states
   fs_storage_reader = dcp.FileSystemReader(checkpoint_path)

   model_state_dict, optimizer_state_dict = get_state_dict(model, optimizer)
   dcp_state_dict = {
       "model": model_state_dict,
       "optimizer": optimizer_state_dict,
   }
   dcp.load(dcp_state_dict, storage_reader=fs_storage_reader)
   assert optimizer_state_dict["param_groups"][0]["lr"] == dcp_state_dict["optimizer"]["param_groups"][0]["lr"]
   set_state_dict(model, optimizer, model_state_dict=dcp_state_dict["model"], optim_state_dict=dcp_state_dict["optimizer"])
   
   # 2. Load global states
   global_state_dict = torch.load(os.path.join(checkpoint_path, "global_state_dict.pt"))
   scheduler.load_state_dict(global_state_dict["scheduler"])
```

**Comment by wz337** [comment]
@Jackmin801 I put up a fix for this. https://github.com/pytorch/pytorch/pull/129398
